drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -679,6 +703,63 @@ def has_submission_deadline_passed(self):
         else:
             return False
 
+    @property
+    def closed(self):
+        """
+        Is the student still allowed to submit answers?
+        """
+        if not self.attempts_remain:
+            return True
+        if self.has_submission_deadline_passed:
+            return True
+
+        return False
+
+    @property
+    def is_attempted(self):
+        """
+        Has the problem been attempted?
+        """
+        return self.attempts > 0
+
+    @property
+    def is_finished(self):
+        """
+        Returns True if answer is closed or answer is correct.
+        """
+        return self.closed or self.is_correct
+
+    @property
+    def is_answer_available(self):
+        """
+        Is student allowed to see an answer?
+        """
+        def _return_false(_x):
+            return False
+        _look_up = {
+            SHOWANSWER.NEVER: lambda _self: False,
+            SHOWANSWER.ATTEMPTED: lambda _self: _self.is_attempted or _self.has_submission_deadline_passed,
+            SHOWANSWER.ANSWERED: lambda _self: _self.is_correct,
+            SHOWANSWER.CLOSED: lambda _self: _self.closed,
+            SHOWANSWER.FINISHED: lambda _self: _self.is_finished,
+            SHOWANSWER.CORRECT_OR_PAST_DUE: lambda _self: _self.is_correct or _self.has_submission_deadline_passed,
+            SHOWANSWER.PAST_DUE: lambda _self: _self.has_submission_deadline_passed,
+            SHOWANSWER.ALWAYS: lambda _self: True,
+            SHOWANSWER.AFTER_ALL_ATTEMPTS: lambda _self: not _self.attempts_remain,
+            SHOWANSWER.AFTER_ALL_ATTEMPTS_OR_CORRECT: lambda _self: not _self.attempts_remain or _self.is_correct,
+            SHOWANSWER.ATTEMPTED_NO_PAST_DUE: lambda _self: _self.is_attempted,
+        }
+
+        if self.mode != Constants.ASSESSMENT_MODE:
+            return False
+        _current_user = self.runtime.service(self, 'user').get_current_user()
+        user_is_staff = _current_user.opt_attrs.get(Constants.ATTR_KEY_USER_IS_STAFF)
+        if self.showanswer not in [SHOWANSWER.NEVER, ''] and user_is_staff:
+            # admins can see the answer unless the problem explicitly prevents it.
+            return True
+        check_available_fun = _look_up.get(self.showanswer, _return_false)
```

Comment:
```suggestion
        check_available_function = _look_up.get(self.showanswer, lambda: True)
```

There is still some `"fun" in "function"`. We don't need this abbreviation.